pkg/data/data.go

```diff
@@ -79,14 +80,13 @@ func (job *Job) Save(db *sqlx.DB) (sql.Result, error) {
 
 func (job *Job) AuthSignature(secret string) string {
 	input := fmt.Sprintf(
-		"%s:%s:%s:%s",
+		"%s:%s:%s",
 		job.ID,
 		job.Email,
 		job.PublishedAt.String(),
-		secret,
 	)
 
-	hash := sha1.Sum([]byte(input))
+	hash := hmac.New(sha256.New, []byte(secret)).Sum([]byte(input))
```

Comment:
Here's a really subtle thing that took me a bit to notice, it has to do with the `.Sum` method on a [`hash.Hash`](https://pkg.go.dev/hash#Hash).

> // Sum appends the current hash to b and returns the resulting slice.
> // It does not change the underlying hash state.
> Sum(b []byte) []byte

The key thing here is that it appends it to b, without any additional hashing. So we can take the resulting hash and just print it out to see the input. Using `Sum` also means the length of the hash is not constant.

Instead, we should use the `Write` method on the hash that is returned from `hmac.New` which would end up looking like:

```go
hash := hmac.New(sha256.New, []byte(secret))
hash.Write([]byte(input))

return base64.URLEncoding.EncodeToString(hash.Sum(nil))
```

Reply:
good catch – i always forget how weird golang is about handling hashes.